### DISCUSSION TITLE
#MVUX: binding does not trigger, when model property updated

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno#16277] [Q&A] [answered]
Hi,

I want to have a state with many properties like:
`record Parameters(string A, string B, string C, ...);`
On view side I want to present a value that depends on some state's properties, like A + B + C.
`new TextBlock().Text(x => x.Bind(() => vm.Parameters).Convert(p => p.A + p.B + p.C)),`
Obviously, I expect that if any of A, B, C is modified, the `TextBlock` above will reflect the change.
Unfortunately it seems not to work:

Have a look.
I simplified the state a bit to produce cleaner code:

This is my model:
```
internal partial record MainModel
{
    // this is the model's state
    public IState<Parameters> Parameters => State.Value(this, () => new Parameters("start"));

    // this method is used to modify the model's state
    public async Task ModifyState()
    {
        await Parameters.UpdateAsync(p => new Parameters(p.A + Random.Shared.Next(10).ToString()));
    }

    // this method is used to check the model's state
    public async Task Check()
    {
        var parameters = await Parameters;
        Debug.WriteLine(parameters.A);
    }
}

public record Parameters(string A);
```

Now, I just bind the state to a view.
Following use-cases are tested:
 - Use-Case 1: TwoWay binding to state's property     -> works fine
 - Use-Case 2: OneWay binding to state's property     -> works fine
 - Use-Case 3: OneWay binding to state with a converter -> **does not work as expected**

Is it a bug or do I am missing something?

```
public sealed partial class MainPage : Page
{
    public MainPage() => this
        .DataContext(new BindableMainModel(), (page, vm) => page
        .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
        .Content(new StackPanel()
            .Orientation(Orientation.Vertical)
            .VerticalAlignment(VerticalAlignment.Center)
            .HorizontalAlignment(HorizontalAlignment.Center)
            .Children(
                // Use-Case 1: TwoWay binding to state's property
                // this binding works correctly
                // - model is updated when the TextBox loses focus.
                // - model is updated correctly, you may check it by clicking the Check button
                new TextBox()
                .Text(x => x.Bind(() => vm.Parameters.A).TwoWay()),

                // Use-Case 2: OneWay binding to state's property
                // this binding works correctly.
                // - when the TextBox loses focus, this TextBlock is updated
                // - when the ModifyState button is clicked, this TextBlock is updated
                // all is OK
                new TextBlock()
                .Text(x => x.Bind(() => vm.Parameters.A)),

                // Use-Case 3: OneWay binding to state with a converter
                // this binding does not work correctly.
                // - when the TextBox loses focus, this TextBlock IS NOT updated
                // - the TextBlock is updated only when the ModifyState button is clicked
                new TextBlock()
                .Text(x => x.Bind(() => vm.Parameters).Convert(p => p.A)),

                // this just adds a random digit to the model's text
                new Button()
                .Content("ModifyState").Command(() => vm.ModifyState),

                // this just prints the model's text to the Output window
                new Button()
                .Content("Check").Command(() => vm.Check)
        )));
```

**Comment by dr1rrb** [comment, marked as answer]
Well indeed it's bug with Uno's binding engine! Opened a new one https://github.com/unoplatform/uno/issues/16297 to track it.

In the `BindableMainModel` we have a property `BindableParameters Parameters { get; }`. When any value of the `Parameters` record is changed, we are updating the internal state of `BindableParameters` but we are not creating a new instance. We then raise a property changed "Parameters" directly on the `BindableMainModel`. With uno's binding engine as the instance of the `BindableParameters` didn't change, the binding engine stops update.

# Workaround
As a workaround, you can use the magic `Value` property: 
`.Text(x => x.Bind(() => vm.Parameters.Value).Convert(p => p.A)),`

**Reply by kucint** [reply]
great!
waiting for the fix then :)
thanks!
shall I close this discussion? or rather wait till the fix is available?
